File: library_third_party_sso_java/src/main/java/com/example/library_third_party_sso_java/GoogleLogin.java
package com.example.library_third_party_sso_java;

import android.app.Activity;
import android.content.Intent;
import android.util.Log;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;

/****************************************************
 * Copyright (C) Alan Corporation. All rights reserved.
 *
 * Author: AlanLai
 * Create Date: 2020-02-17
 * Usage:
 *
 * Revision History
 * Date         Author           Description
 ****************************************************/

class GoogleLogin {
    private static final int RC_SIGN_IN = 1293;
    private Activity mActivity;
    private ThirdPartySSOCallback mThirdPartySSOCallback;
    private GoogleSignInOptions mGoogleSignInOptions;
    private GoogleSignInClient mGoogleSignInClient;
    private GoogleSignInAccount mGoogleSignInAccount;

    public GoogleLogin(Activity activity) {
        mActivity = activity;

        mGoogleSignInOptions = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_GAMES_SIGN_IN)
                .requestEmail().build();

        mGoogleSignInClient = GoogleSignIn.getClient(mActivity, mGoogleSignInOptions);
    }

    public void onStart() {
        mGoogleSignInAccount = GoogleSignIn.getLastSignedInAccount(mActivity);
    }

    public void setThirdPartySSOCallback(ThirdPartySSOCallback thirdPartySSOCallback) {
        mThirdPartySSOCallback = thirdPartySSOCallback;
    }

    public void signIn() {
        Intent signInIntent = mGoogleSignInClient.getSignInIntent();
        mActivity.startActivityForResult(signInIntent, RC_SIGN_IN);

    }

    public void onActivityResult(int requestCode, int resultCode, Intent intent) {
        if (requestCode == RC_SIGN_IN) {
            // The Task returned from this call is always completed, no need to attach
            // a listener.
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(intent);
            handleSignInResult(task);
        }
    }

    private void handleSignInResult(Task<GoogleSignInAccount> completedTask) {
        try {
            mGoogleSignInAccount = completedTask.getResult(ApiException.class);

            // Signed in successfully, show authenticated UI.

            GoogleData googleData = new GoogleData();

            if (mGoogleSignInAccount != null) {
                googleData.setEmail(mGoogleSignInAccount.getEmail());
            }


            mThirdPartySSOCallback.updateGoogleUI(googleData);
        } catch (ApiException e) {
            // The ApiException status code indicates the detailed failure reason.
            // Please refer to the GoogleSignInStatusCodes class reference for more information.
            Log.e("GoogleLogin", "signInResult:failed code=" + e.getStatusCode());
            mThirdPartySSOCallback.updateGoogleUI(null);
        }
    }
}

File: library_third_party_sso_kotlin/src/main/java/com/example/library_third_party_sso_kotlin/GoogleLogin.kt
package com.example.library_third_party_sso_kotlin

import android.content.Intent
import android.util.Log
import android.widget.Toast
import androidx.fragment.app.FragmentActivity
import com.example.library_third_party_login.R
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.common.api.ApiException


internal class GoogleLogin(private var activity: FragmentActivity, private val resultCallback: LoginResultCallback) : IThirdPartyLogin {
    private var googleSignInClient: GoogleSignInClient? = null

    companion object {
        const val RC_SIGN_IN = 99
    }

    init {
        Log.e("GoogleLogin", "google_login_client_id: ${activity.getString(R.string.google_login_client_id)}")
        if (checkLocalPropertiesIsNotEmpty()) {
            // Configure sign-in to request the user's ID, email address, and basic
            // profile. ID and basic profile are included in DEFAULT_SIGN_IN.
            val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                    .requestEmail()
                    .requestIdToken(activity.getString(R.string.google_login_client_id))
                    .build()
            // Build a GoogleSignInClient with the options specified by gso.
            googleSignInClient = GoogleSignIn.getClient(activity, gso)
        }
    }

    override fun logIn() {
        if (checkLocalPropertiesIsNotEmpty()) {
            //        if (!isLogin())
            activity.startActivityForResult(googleSignInClient?.signInIntent, RC_SIGN_IN)

        }
    }

    override fun logOut() {
        if (checkLocalPropertiesIsNotEmpty()) {
            if (isLogin())
                googleSignInClient?.signOut()
        }
    }

    override fun isLogin(): Boolean {
        return if (checkLocalPropertiesIsNotEmpty()) {
            // Check for existing Google Sign In account, if the user is already signed in
            // the GoogleSignInAccount will be non-null.
            val account = GoogleSignIn.getLastSignedInAccount(activity)
            Log.e("GoogleLogin", "isLogin: ${account != null}")
            Log.e("GoogleLogin", account?.email + "/" +
                    account?.displayName + "/" +
                    account?.id + "/" +
                    account?.idToken)
            return account != null
        } else {
            false
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (checkLocalPropertiesIsNotEmpty()) {
            // Result returned from launching the Intent from GoogleSignInClient.getSignInIntent(...);
            if (requestCode == RC_SIGN_IN) {
                // The Task returned from this call is always completed, no need to attach
                // a listener.
                val task = GoogleSignIn.getSignedInAccountFromIntent(data)

                try {
                    val account = task.getResult(ApiException::class.java)
                    // Signed in successfully, show authenticated UI.
                    Log.e("GoogleLogin", account?.idToken)
                    resultCallback.onSuccess(account?.email + "/" +
                            account?.displayName + "/" +
                            account?.id + "/" +
                            account?.idToken)
                } catch (e: ApiException) {
                    // The ApiException status code indicates the detailed failure reason.
                    // Please refer to the GoogleSignInStatusCodes class reference for more information.
                    Log.e("GoogleLogin", "signInResult:failed code: " + e.statusCode)
                    Toast.makeText(activity, "Please refer to the GoogleSignInStatusCodes class reference for more information => code: ${e.statusCode}", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    private fun checkLocalPropertiesIsNotEmpty(): Boolean {
        return if (activity.getString(R.string.google_login_client_id).isNotEmpty()) {
            true
        } else {
            Toast.makeText(activity, "google_login_client_id is empty", Toast.LENGTH_SHORT).show()
            false
        }
    }

}